File: src/platform/mac/window.m
#import <Cocoa/Cocoa.h>
#import <Metal/Metal.h>
#import <MetalKit/MetalKit.h>
#import "platform/window.h"
#import "platform/renderer.h"
#import "platform/mac/app_delegate.h"
#import "platform/mac/renderer_private.h"
#import "platform/renderer_private.h"
#import "core/memory_manager.h"
#import "types/input_state.h"
#import "types/input_state_private.h"

/**
 * @brief Platform-specific window structure for macOS Metal implementation.
 * 
 * @details This structure wraps the macOS NSWindow and NSView handles along
 *          with the engine's input state for the Metal platform. It provides
 *          the bridge between macOS window events and the engine's input system.
 */
typedef struct EseMetalWindow {
    NSWindow *window;              /**< macOS NSWindow handle */
    NSView *view;                  /**< macOS NSView handle for the window */
    EseInputState *inputState;     /**< Reference to the engine's input state */
} EseMetalWindow;

// Map macOS keycodes to your engine's keys
static EseInputKey _mapMacOSKeycodeToInputKey(unsigned short keyCode) {
    switch (keyCode) {
        case 0x00: return InputKey_A;
        case 0x0B: return InputKey_B;
        case 0x08: return InputKey_C;
        case 0x02: return InputKey_D;
        case 0x0E: return InputKey_E;
        case 0x03: return InputKey_F;
        case 0x05: return InputKey_G;
        case 0x04: return InputKey_H;
        case 0x22: return InputKey_I;
        case 0x26: return InputKey_J;
        case 0x28: return InputKey_K;
        case 0x25: return InputKey_L;
        case 0x2E: return InputKey_M;
        case 0x2D: return InputKey_N;
        case 0x1F: return InputKey_O;
        case 0x23: return InputKey_P;
        case 0x0C: return InputKey_Q;
        case 0x0F: return InputKey_R;
        case 0x01: return InputKey_S;
        case 0x11: return InputKey_T;
        case 0x20: return InputKey_U;
        case 0x09: return InputKey_V;
        case 0x0D: return InputKey_W;
        case 0x07: return InputKey_X;
        case 0x10: return InputKey_Y;
        case 0x06: return InputKey_Z;
        
        // Numbers
        case 0x1D: return InputKey_0;
        case 0x12: return InputKey_1;
        case 0x13: return InputKey_2;
        case 0x14: return InputKey_3;
        case 0x15: return InputKey_4;
        case 0x17: return InputKey_5;
        case 0x16: return InputKey_6;
        case 0x1A: return InputKey_7;
        case 0x1C: return InputKey_8;
        case 0x19: return InputKey_9;
        
        // Function keys
        case 0x7A: return InputKey_F1;
        case 0x78: return InputKey_F2;
        case 0x63: return InputKey_F3;
        case 0x76: return InputKey_F4;
        case 0x60: return InputKey_F5;
        case 0x61: return InputKey_F6;
        case 0x62: return InputKey_F7;
        case 0x64: return InputKey_F8;
        case 0x65: return InputKey_F9;
        case 0x6D: return InputKey_F10;
        case 0x67: return InputKey_F11;
        case 0x6F: return InputKey_F12;
        
        // Arrow keys
        case 0x7E: return InputKey_UP;
        case 0x7D: return InputKey_DOWN;
        case 0x7B: return InputKey_LEFT;
        case 0x7C: return InputKey_RIGHT;
        
        // Special keys
        case 0x31: return InputKey_SPACE;
        case 0x24: return InputKey_ENTER;
        case 0x35: return InputKey_ESCAPE;
        case 0x30: return InputKey_TAB;
        case 0x33: return InputKey_BACKSPACE;
        case 0x39: return InputKey_CAPSLOCK;
        
        default: return InputKey_UNKNOWN;

    }
}

EseWindow* window_create(int width, int height, const char* title) {
    if (NSApp == nil) {
        [NSApplication sharedApplication];
    }

    // Setup menu bar with Quit
    NSMenu *menubar = [[NSMenu alloc] init];
    [NSApp setMainMenu:menubar];

    NSMenuItem *appMenuItem = [[NSMenuItem alloc] init];
    [menubar addItem:appMenuItem];

    NSMenu *appMenu = [[NSMenu alloc] init];
    [appMenuItem setSubmenu:appMenu];

    NSString *appName = [[NSProcessInfo processInfo] processName];
    NSString *quitTitle = [NSString stringWithFormat:@"Quit %@", appName];
    NSMenuItem *quitItem = [[NSMenuItem alloc]
        initWithTitle:quitTitle
               action:@selector(terminate:)
        keyEquivalent:@"q"];
    [appMenu addItem:quitItem];

    // Create window + engine structs
    EseWindow* window = (EseWindow*)memory_manager.malloc(sizeof(EseWindow), MMTAG_WINDOW);
    EseMetalWindow* metalWindow = (EseMetalWindow*)memory_manager.malloc(sizeof(EseMetalWindow), MMTAG_WINDOW);
    metalWindow->inputState = ese_input_state_create(nil);

    // Create window
    NSRect frame = NSMakeRect(100, 100, width, height);
    metalWindow->window = [[NSWindow alloc] initWithContentRect:frame
                                                      styleMask:(NSWindowStyleMaskTitled |
                                                                 NSWindowStyleMaskClosable |
                                                                 NSWindowStyleMaskResizable)
                                                        backing:NSBackingStoreBuffered
                                                          defer:NO];
    metalWindow->window.title = [NSString stringWithUTF8String:title];

    // Activate app and show window
    [NSApp activateIgnoringOtherApps:YES];
    [metalWindow->window makeKeyAndOrderFront:nil];
    [metalWindow->window setAcceptsMouseMovedEvents:YES];

    // Hook delegate
    EseAppDelegate *delegate = [[EseAppDelegate alloc] init];
    delegate.eseWindow = window;
    [NSApp setDelegate:delegate];

    window->platform_window = metalWindow;
    window->width = width;
    window->height = height;
    window->should_close = false;

    return window;
}

void window_destroy(EseWindow* window) {
    if (!window) return;
    if (window->platform_window) {
        EseMetalWindow* metalWindow = (EseMetalWindow*)window->platform_window;
        [metalWindow->window close];

        ese_input_state_destroy(metalWindow->inputState);
        memory_manager.free(metalWindow);
    }
    memory_manager.free(window);
}

void window_set_renderer(EseWindow* window, EseRenderer* renderer) {
    if (!window || !renderer) return;
    EseMetalWindow* metalWindow = (EseMetalWindow*)window->platform_window;

    EseMetalRenderer *internal = (EseMetalRenderer *)renderer->internal;
    window->renderer = renderer;

    // Set the renderer view
    metalWindow->view = internal->view;
    [metalWindow->window.contentView addSubview:metalWindow->view];
    metalWindow->view.frame = metalWindow->window.contentView.bounds;
    metalWindow->view.autoresizingMask = NSViewWidthSizable | NSViewHeightSizable;

    // Enforce hiDPI or lowDPI after attaching to window
    if (!renderer->hiDPI) {
        // Force 1:1 pixel mapping
        internal->view.layer.contentsScale = 1.0;
        internal->view.drawableSize = internal->view.bounds.size;
    } else {
        // Match screen scale (Retina)
        CGFloat scale = metalWindow->window.screen.backingScaleFactor;
        internal->view.layer.contentsScale = scale;
        CGSize pointSize = internal->view.bounds.size;
        internal->view.drawableSize = CGSizeMake(pointSize.width * scale,
                                                           pointSize.height * scale);
    }
}

void window_process(EseWindow* window, EseInputState* out_input_state) {
    if (!window) return;

    EseMetalWindow* metalWindow = (EseMetalWindow*)window->platform_window;
    if (!metalWindow) return;

    if (window->should_close) {
        [metalWindow->window close];
    }

    NSEvent *event;
    while ((event = [NSApp nextEventMatchingMask:NSEventMaskAny
                                       untilDate:[NSDate distantPast]
                                          inMode:NSDefaultRunLoopMode
                                         dequeue:YES])) {

        // Menu handled it; skip engine handling and don't cause beep.
        if ([[NSApp mainMenu] performKeyEquivalent:event]) {
            continue;
        }

        // --- Update modifier keys for every event ---
        NSEventModifierFlags mods = [event modifierFlags];
        metalWindow->inputState->keys_down[InputKey_LSHIFT]  = (mods & NSEventModifierFlagShift) != 0;
        metalWindow->inputState->keys_down[InputKey_RSHIFT]  = (mods & NSEventModifierFlagShift) != 0;

        metalWindow->inputState->keys_down[InputKey_LCTRL]   = (mods & NSEventModifierFlagControl) != 0;
        metalWindow->inputState->keys_down[InputKey_RCTRL]   = (mods & NSEventModifierFlagControl) != 0;

        metalWindow->inputState->keys_down[InputKey_LALT]    = (mods & NSEventModifierFlagOption) != 0;
        metalWindow->inputState->keys_down[InputKey_RALT]    = (mods & NSEventModifierFlagOption) != 0;

        metalWindow->inputState->keys_down[InputKey_LCMD]    = (mods & NSEventModifierFlagCommand) != 0;
        metalWindow->inputState->keys_down[InputKey_RCMD]    = (mods & NSEventModifierFlagCommand) != 0;

        metalWindow->inputState->keys_down[InputKey_CAPSLOCK] = (mods & NSEventModifierFlagCapsLock) != 0;

        switch ([event type]) {
            case NSEventTypeKeyDown: {
                EseInputKey key = _mapMacOSKeycodeToInputKey([event keyCode]);
                if (key != InputKey_UNKNOWN && key < InputKey_MAX) {
                    metalWindow->inputState->keys_down[key] = true;
                    metalWindow->inputState->keys_pressed[key] = true;
                }
                // Dont forward the event to prevent beeps
                continue;
            } break;
            case NSEventTypeKeyUp: {
                EseInputKey key = _mapMacOSKeycodeToInputKey([event keyCode]);
                if (key != InputKey_UNKNOWN && key < InputKey_MAX) {
                    metalWindow->inputState->keys_down[key] = false;
                    metalWindow->inputState->keys_released[key] = true;
                }
            } break;
            case NSEventTypeLeftMouseDown:
                metalWindow->inputState->mouse_buttons[0] = true;
                break;
            case NSEventTypeLeftMouseUp:
                metalWindow->inputState->mouse_buttons[0] = false;
                break;
            case NSEventTypeRightMouseDown:
                metalWindow->inputState->mouse_buttons[1] = true;
                break;
            case NSEventTypeRightMouseUp:
                metalWindow->inputState->mouse_buttons[1] = false;
                break;
            case NSEventTypeMouseMoved:
            case NSEventTypeLeftMouseDragged:
            case NSEventTypeRightMouseDragged: {
                NSPoint loc = [metalWindow->view convertPoint:[event locationInWindow] fromView:nil];
                metalWindow->inputState->mouse_x = loc.x;
                metalWindow->inputState->mouse_y = loc.y;
            } break;
            case NSEventTypeScrollWheel:
                metalWindow->inputState->mouse_scroll_dx += [event scrollingDeltaX];
                metalWindow->inputState->mouse_scroll_dy += [event scrollingDeltaY];
                break;
            default:
                break;
        }
        [NSApp sendEvent:event];
    }

    // Copy input state to output
    size_t prefix = offsetof(EseInputState, state);
    memmove(out_input_state, metalWindow->inputState, prefix);

    // Reset per-frame input
    memset(metalWindow->inputState->keys_pressed, 0, sizeof(metalWindow->inputState->keys_pressed));
    memset(metalWindow->inputState->keys_released, 0, sizeof(metalWindow->inputState->keys_released));
    metalWindow->inputState->mouse_scroll_dx = 0;
    metalWindow->inputState->mouse_scroll_dy = 0;

    if (!metalWindow->window || ![metalWindow->window isVisible]) {
        window->should_close = true;
        return;
    }

    [metalWindow->view setNeedsDisplay:YES];
}

void window_close(EseWindow* window) {
    if (!window) return;
    EseMetalWindow* metalWindow = (EseMetalWindow*)window->platform_window;
    if (metalWindow && metalWindow->window) {
        [metalWindow->window close]; // actually close the macOS window
    }
}

bool window_should_close(EseWindow* window) {
    if (!window) return true;
    return window->should_close;
}

File: src/platform/glfw/window.c
#include "platform/window.h"
#include "core/memory_manager.h"
#include "platform/glfw/renderer_private.h"
#include "platform/renderer.h"
#include "platform/renderer_private.h"
#include "types/input_state.h"
#include "types/input_state_private.h"
#include "utility/log.h"
#include <GL/glew.h>
#include <GLFW/glfw3.h>
#include <stddef.h>
#include <stdio.h>
#include <string.h>

/**
 * @brief Platform-specific window structure for GLFW implementation.
 *
 * @details This structure wraps the GLFW window handle and input state
 *          for the GLFW platform. It provides the bridge between GLFW
 *          window events and the engine's input system.
 */
typedef struct EseGLFWWindow {
  GLFWwindow *glfw_window;   /** GLFW window handle */
  EseInputState *inputState; /** Reference to the engine's input state */
} EseGLFWWindow;

// Map GLFW keycodes to your engine's keys
static EseInputKey mapGLFWKeyToInputKey(int key) {
  switch (key) {
  case GLFW_KEY_A:
    return InputKey_A;
  case GLFW_KEY_B:
    return InputKey_B;
  case GLFW_KEY_C:
    return InputKey_C;
  case GLFW_KEY_D:
    return InputKey_D;
  case GLFW_KEY_E:
    return InputKey_E;
  case GLFW_KEY_F:
    return InputKey_F;
  case GLFW_KEY_G:
    return InputKey_G;
  case GLFW_KEY_H:
    return InputKey_H;
  case GLFW_KEY_I:
    return InputKey_I;
  case GLFW_KEY_J:
    return InputKey_J;
  case GLFW_KEY_K:
    return InputKey_K;
  case GLFW_KEY_L:
    return InputKey_L;
  case GLFW_KEY_M:
    return InputKey_M;
  case GLFW_KEY_N:
    return InputKey_N;
  case GLFW_KEY_O:
    return InputKey_O;
  case GLFW_KEY_P:
    return InputKey_P;
  case GLFW_KEY_Q:
    return InputKey_Q;
  case GLFW_KEY_R:
    return InputKey_R;
  case GLFW_KEY_S:
    return InputKey_S;
  case GLFW_KEY_T:
    return InputKey_T;
  case GLFW_KEY_U:
    return InputKey_U;
  case GLFW_KEY_V:
    return InputKey_V;
  case GLFW_KEY_W:
    return InputKey_W;
  case GLFW_KEY_X:
    return InputKey_X;
  case GLFW_KEY_Y:
    return InputKey_Y;
  case GLFW_KEY_Z:
    return InputKey_Z;

  case GLFW_KEY_0:
    return InputKey_0;
  case GLFW_KEY_1:
    return InputKey_1;
  case GLFW_KEY_2:
    return InputKey_2;
  case GLFW_KEY_3:
    return InputKey_3;
  case GLFW_KEY_4:
    return InputKey_4;
  case GLFW_KEY_5:
    return InputKey_5;
  case GLFW_KEY_6:
    return InputKey_6;
  case GLFW_KEY_7:
    return InputKey_7;
  case GLFW_KEY_8:
    return InputKey_8;
  case GLFW_KEY_9:
    return InputKey_9;

  case GLFW_KEY_F1:
    return InputKey_F1;
  case GLFW_KEY_F2:
    return InputKey_F2;
  case GLFW_KEY_F3:
    return InputKey_F3;
  case GLFW_KEY_F4:
    return InputKey_F4;
  case GLFW_KEY_F5:
    return InputKey_F5;
  case GLFW_KEY_F6:
    return InputKey_F6;
  case GLFW_KEY_F7:
    return InputKey_F7;
  case GLFW_KEY_F8:
    return InputKey_F8;
  case GLFW_KEY_F9:
    return InputKey_F9;
  case GLFW_KEY_F10:
    return InputKey_F10;
  case GLFW_KEY_F11:
    return InputKey_F11;
  case GLFW_KEY_F12:
    return InputKey_F12;

  case GLFW_KEY_UP:
    return InputKey_UP;
  case GLFW_KEY_DOWN:
    return InputKey_DOWN;
  case GLFW_KEY_LEFT:
    return InputKey_LEFT;
  case GLFW_KEY_RIGHT:
    return InputKey_RIGHT;

  case GLFW_KEY_SPACE:
    return InputKey_SPACE;
  case GLFW_KEY_ENTER:
    return InputKey_ENTER;
  case GLFW_KEY_ESCAPE:
    return InputKey_ESCAPE;
  case GLFW_KEY_TAB:
    return InputKey_TAB;
  case GLFW_KEY_BACKSPACE:
    return InputKey_BACKSPACE;
  case GLFW_KEY_CAPS_LOCK:
    return InputKey_CAPSLOCK;

  default:
    return InputKey_UNKNOWN;
  }
}

// Callback: key events
static void glfw_key_callback(GLFWwindow *w, int key, int scancode, int action,
                              int mods) {
  EseGLFWWindow *pw = (EseGLFWWindow *)glfwGetWindowUserPointer(w);
  if (!pw || !pw->inputState)
    return;

  // update modifiers
  pw->inputState->keys_down[InputKey_LSHIFT] = (mods & GLFW_MOD_SHIFT) != 0;
  pw->inputState->keys_down[InputKey_RSHIFT] = (mods & GLFW_MOD_SHIFT) != 0;

  pw->inputState->keys_down[InputKey_LCTRL] = (mods & GLFW_MOD_CONTROL) != 0;
  pw->inputState->keys_down[InputKey_RCTRL] = (mods & GLFW_MOD_CONTROL) != 0;

  pw->inputState->keys_down[InputKey_LALT] = (mods & GLFW_MOD_ALT) != 0;
  pw->inputState->keys_down[InputKey_RALT] = (mods & GLFW_MOD_ALT) != 0;

  pw->inputState->keys_down[InputKey_LCMD] = (mods & GLFW_MOD_SUPER) != 0;
  pw->inputState->keys_down[InputKey_RCMD] = (mods & GLFW_MOD_SUPER) != 0;

  pw->inputState->keys_down[InputKey_CAPSLOCK] =
      (mods & GLFW_MOD_CAPS_LOCK) != 0;

  if (key == GLFW_KEY_UNKNOWN)
    return;

  EseInputKey ik = mapGLFWKeyToInputKey(key);
  if (ik == InputKey_UNKNOWN || ik >= InputKey_MAX)
    return;

  if (action == GLFW_PRESS) {
    pw->inputState->keys_down[ik] = true;
    pw->inputState->keys_pressed[ik] = true;
  } else if (action == GLFW_RELEASE) {
    pw->inputState->keys_down[ik] = false;
    pw->inputState->keys_released[ik] = true;
  }
}

// Callback: mouse buttons
static void glfw_mouse_button_callback(GLFWwindow *w, int button, int action,
                                       int mods) {
  EseGLFWWindow *pw = (EseGLFWWindow *)glfwGetWindowUserPointer(w);
  if (!pw || !pw->inputState)
    return;

  int i = -1;
  if (button == GLFW_MOUSE_BUTTON_LEFT) {
    i = InputMouse_LEFT;
  } else if (button == GLFW_MOUSE_BUTTON_RIGHT) {
    i = InputMouse_RIGHT;
  } else if (button == GLFW_MOUSE_BUTTON_MIDDLE) {
    i = InputMouse_MIDDLE;
  } else if (button == GLFW_MOUSE_BUTTON_X1) {
    i = InputMouse_X1;
  } else if (button == GLFW_MOUSE_BUTTON_X2) {
    i = InputMouse_X2;
  }

  if (i == -1)
    return;

  if (action == GLFW_PRESS) {
    if (!pw->inputState->mouse_down[i]) {
      pw->inputState->mouse_clicked[i] = true; // first down edge
    }
    pw->inputState->mouse_down[i] = true;
  } else if (action == GLFW_RELEASE) {
    pw->inputState->mouse_down[i] = false;
    pw->inputState->mouse_released[i] = true; // release edge
  }
}

// Callback: cursor position
static void glfw_cursor_pos_callback(GLFWwindow *w, double xpos, double ypos) {
  EseGLFWWindow *pw = (EseGLFWWindow *)glfwGetWindowUserPointer(w);
  if (!pw || !pw->inputState)
    return;

  pw->inputState->mouse_x = (float)xpos;
  pw->inputState->mouse_y = (float)ypos;
}

// Callback: scroll
static void glfw_scroll_callback(GLFWwindow *w, double xoffset,
                                 double yoffset) {
  EseGLFWWindow *pw = (EseGLFWWindow *)glfwGetWindowUserPointer(w);
  if (!pw || !pw->inputState)
    return;

  pw->inputState->mouse_scroll_dx += (float)xoffset;
  pw->inputState->mouse_scroll_dy += (float)yoffset;
}

// Callback: window close
static void glfw_window_close_callback(GLFWwindow *w) {
  EseGLFWWindow *pw = (EseGLFWWindow *)glfwGetWindowUserPointer(w);
  if (!pw)
    return;

  if (pw->glfw_window) {
    glfwSetWindowShouldClose(pw->glfw_window, GLFW_TRUE);
  }
}

EseWindow *window_create(int width, int height, const char *title) {
  if (!glfwInit()) {
    log_error("Failed to initialize GLFW\n");
    return NULL;
  }

  // OpenGL hints (keep compatible with previous file)
  glfwWindowHint(GLFW_CONTEXT_VERSION_MAJOR, 3);
  glfwWindowHint(GLFW_CONTEXT_VERSION_MINOR, 3);
  glfwWindowHint(GLFW_OPENGL_PROFILE, GLFW_OPENGL_CORE_PROFILE);
#ifdef __APPLE__
  glfwWindowHint(GLFW_OPENGL_FORWARD_COMPAT, GL_TRUE);
#endif

  GLFWwindow *glfwWin =
      glfwCreateWindow(width, height, title ? title : "EseWindow", NULL, NULL);
  if (!glfwWin) {
    fprintf(stderr, "Failed to create GLFW window\n");
    glfwTerminate();
    return NULL;
  }

  glfwMakeContextCurrent(glfwWin);
  glfwSwapInterval(1);

  glewExperimental = GL_TRUE;
  GLenum err = glewInit();
  if (err != GLEW_OK) {
    fprintf(stderr, "Failed to initialize GLEW: %s\n", glewGetErrorString(err));
    glfwDestroyWindow(glfwWin);
    glfwTerminate();
    return NULL;
  }

  EseWindow *win =
      (EseWindow *)memory_manager.malloc(sizeof(EseWindow), MMTAG_WINDOW);
  memset(win, 0, sizeof(EseWindow));
  EseGLFWWindow *pw = (EseGLFWWindow *)memory_manager.malloc(
      sizeof(EseGLFWWindow), MMTAG_WINDOW);
  memset(pw, 0, sizeof(EseGLFWWindow));

  pw->glfw_window = glfwWin;
  // create input state
  pw->inputState = ese_input_state_create(NULL);

  // attach user pointer and callbacks
  glfwSetWindowUserPointer(glfwWin, pw);
  glfwSetKeyCallback(glfwWin, glfw_key_callback);
  glfwSetMouseButtonCallback(glfwWin, glfw_mouse_button_callback);
  glfwSetCursorPosCallback(glfwWin, glfw_cursor_pos_callback);
  glfwSetScrollCallback(glfwWin, glfw_scroll_callback);
  glfwSetWindowCloseCallback(glfwWin, glfw_window_close_callback);

  win->platform_window = pw;
  win->width = width;
  win->height = height;
  win->should_close = false;

  return win;
}

void window_destroy(EseWindow *window) {
  if (!window)
    return;
  if (window->platform_window) {
    EseGLFWWindow *pw = (EseGLFWWindow *)window->platform_window;
    if (pw->glfw_window) {
      glfwDestroyWindow(pw->glfw_window);
      pw->glfw_window = NULL;
    }
    if (pw->inputState) {
      ese_input_state_destroy(pw->inputState);
      pw->inputState = NULL;
    }
    memory_manager.free(pw);
  }
  glfwTerminate();
  memory_manager.free(window);
}

void window_set_renderer(EseWindow *window, EseRenderer *renderer) {
  if (!window)
    return;
  EseGLFWWindow *pw = (EseGLFWWindow *)window->platform_window;
  if (!pw || !pw->glfw_window)
    return;

  window->renderer = renderer;

  if (!renderer) {
    return;
  }

  // Make context current for any GL calls
  glfwMakeContextCurrent(pw->glfw_window);

  if (renderer->hiDPI) {
    // Get the actual framebuffer size (pixels) immediately after window
    // creation
    int framebuffer_width, framebuffer_height;
    glfwGetFramebufferSize(pw->glfw_window, &framebuffer_width,
                           &framebuffer_height);

    // Set the renderer dimensions to the actual framebuffer size
    renderer->view_w = framebuffer_width;
    renderer->view_h = framebuffer_height;

    // Set OpenGL viewport to match
    glViewport(0, 0, framebuffer_width, framebuffer_height);
  } else {
    renderer->view_w = window->width;
    renderer->view_h = window->height;
  }

  // store glfw window in renderer internal if present
  if (renderer->internal) {
    EseGLRenderer *internal = (EseGLRenderer *)renderer->internal;
    internal->window = pw->glfw_window;
  }
}

void window_process(EseWindow *window, EseInputState *out_input_state) {
  if (!window || !out_input_state)
    return;
  EseGLFWWindow *pw = (EseGLFWWindow *)window->platform_window;
  if (!pw || !pw->glfw_window) {
    // Cleanup input state
    memset(pw->inputState->keys_pressed, 0,
           sizeof(pw->inputState->keys_pressed));
    memset(pw->inputState->keys_down, 0, sizeof(pw->inputState->keys_down));
    memset(pw->inputState->keys_released, 0,
           sizeof(pw->inputState->keys_released));
    pw->inputState->mouse_scroll_dx = 0;
    pw->inputState->mouse_scroll_dy = 0;
    memset(pw->inputState->mouse_down, 0, sizeof(pw->inputState->mouse_down));
    memset(pw->inputState->mouse_clicked, 0,
           sizeof(pw->inputState->mouse_clicked));
    memset(pw->inputState->mouse_released, 0,
           sizeof(pw->inputState->mouse_released));

    // copy prefix of input state like macOS code
    size_t prefix = offsetof(EseInputState, state);
    memmove(out_input_state, pw->inputState, prefix);

    return;
  }

  // Poll and swap
  glfwPollEvents();
  glfwMakeContextCurrent(pw->glfw_window);
  glfwSwapBuffers(pw->glfw_window);

  // copy prefix of input state like macOS code
  size_t prefix = offsetof(EseInputState, state);
  memmove(out_input_state, pw->inputState, prefix);

  // reset per-frame data
  memset(pw->inputState->keys_pressed, 0, sizeof(pw->inputState->keys_pressed));
  memset(pw->inputState->keys_released, 0,
         sizeof(pw->inputState->keys_released));
  pw->inputState->mouse_scroll_dx = 0;
  pw->inputState->mouse_scroll_dy = 0;
  memset(pw->inputState->mouse_clicked, 0,
         sizeof(pw->inputState->mouse_clicked));
  memset(pw->inputState->mouse_released, 0,
         sizeof(pw->inputState->mouse_released));

  // Handle window close
  window->should_close = glfwWindowShouldClose(pw->glfw_window);
  if (window->should_close) {
    glfwSetWindowShouldClose(pw->glfw_window, GLFW_TRUE);
    glfwPostEmptyEvent();
    glfwDestroyWindow(pw->glfw_window);
    pw->glfw_window = NULL;
    return;
  }

  if (window->renderer) {
    renderer_draw(window->renderer);
  }
}

void window_close(EseWindow *window) {
  if (!window)
    return;
  EseGLFWWindow *pw = (EseGLFWWindow *)window->platform_window;

  if (pw->glfw_window) {
    glfwSetWindowShouldClose(pw->glfw_window, GLFW_TRUE);
  }

  window->should_close = true;
}

bool window_should_close(EseWindow *window) {
  if (!window)
    return true;
  return window->should_close;
}
